File: SuperiorPackGroup/BLL/UserPermissionsBLL.vb
Imports System
Imports SuperiorPackGroup.SPGTableAdapters
Imports System.Text
Imports DXDAL.SPGData

<System.ComponentModel.DataObject()> _
Public Class UserPermissionsBLL

    Private m_UserPermissionsTableAdapter As UserPermissionsTableAdapter = Nothing
    Public ReadOnly Property Adapter() As UserPermissionsTableAdapter

        Get
            If m_UserPermissionsTableAdapter Is Nothing Then
                m_UserPermissionsTableAdapter = New UserPermissionsTableAdapter()
                m_UserPermissionsTableAdapter.Connection.ConnectionString = My.Settings.UserConnectionString
            End If
            Return m_UserPermissionsTableAdapter
        End Get

    End Property

    <System.ComponentModel.DataObjectMethod(System.ComponentModel.DataObjectMethodType.Select, False)> _
    Public Function GetUserPermissionsByUserName(ByVal userName As String, ByVal permissionObject As String) As SPG.UserPermissionsDataTable

        Return Adapter.GetUserPermissionsByUserName(userName, permissionObject)

    End Function

    Private Sub UpdateAuditTrail(ByVal ModifiedRecord As SPG.UserPermissionsRow, ByVal originalRecord As Object())

        Dim builder As New StringBuilder(String.Empty)
        Dim recordFields As Integer = (ModifiedRecord.ItemArray.Length - 1)
        Dim i As Integer = 0
        Do While (i <= recordFields)
            Try
                If IsDBNull(originalRecord(i)) Then
                    If Not IsDBNull(ModifiedRecord.Item(i)) Then
                        builder.Append(String.Format("{0}:{1}({2}); ", ModifiedRecord.Table.Columns.Item(i).ColumnName, "NULL", ModifiedRecord.Item(i)))
                    End If
                ElseIf Information.IsDBNull(ModifiedRecord.Item(i)) Then
                    If Not Information.IsDBNull(originalRecord(i)) Then
                        builder.Append(String.Format("{0}:{1}({2}); ", ModifiedRecord.Table.Columns.Item(i).ColumnName, originalRecord(i), "NULL"))
                    End If
                ElseIf ModifiedRecord.Item(i) IsNot originalRecord(i) Then
                    builder.Append(String.Format("{0}:{1}({2}); ", ModifiedRecord.Table.Columns.Item(i).ColumnName, originalRecord(i), ModifiedRecord.Item(i)))
                End If
            Catch
            End Try
            i += 1
        Loop
        If (builder.Length > 2) Then
            builder.Length = (builder.Length - 2)
            AuditTrailBLL.AddTrailEntry(My.Settings.UserName, CType(ModifiedRecord.Item(0), Integer), "UserPermissions", builder.ToString)
        End If
    End Sub

    <System.ComponentModel.DataObjectMethod(System.ComponentModel.DataObjectMethodType.Update, True)> _
    Public Function UpdateUserPermissions(ByVal UserPermissionID As Nullable(Of Integer), ByVal PermissionObject As String, ByVal PermissionLevel As String, ByVal UserName As String) As Boolean

        Dim permissions As SPG.UserPermissionsDataTable = Adapter.GetUserPermissionByID(UserPermissionID.Value)

        If permissions.Count = 0 Then
            'It is a new record
            Return Me.InsertUserPermissions(PermissionObject, PermissionLevel, UserName)
        End If

        Dim userPermissions As SPG.UserPermissionsRow = permissions(0)

        Dim originalRecord As Object() = Nothing
        originalRecord = userPermissions.ItemArray

        userPermissions.PermissionObject = PermissionObject
        userPermissions.PermissionLevel = PermissionLevel
        userPermissions.UserName = UserName

        Dim rowsAffected As Integer = Adapter.Update(userPermissions)

        If Not IsNothing(originalRecord) Then
            Me.UpdateAuditTrail(userPermissions, originalRecord)
        End If

        Return rowsAffected = 1

    End Function

    <System.ComponentModel.DataObjectMethod(System.ComponentModel.DataObjectMethodType.Insert, True)> _
    Public Function InsertUserPermissions(ByVal PermissionObject As String, ByVal PermissionLevel As String, ByVal UserName As String) As Boolean

        Dim permissions As SPG.UserPermissionsDataTable = New SPG.UserPermissionsDataTable
        Dim userPermissions As SPG.UserPermissionsRow = permissions.NewUserPermissionsRow

        userPermissions.UserName = UserName
        userPermissions.PermissionObject = PermissionObject
        userPermissions.PermissionLevel = PermissionLevel

        permissions.AddUserPermissionsRow(userPermissions)
        Dim rowsAffected As Integer = Adapter.Update(permissions)

        Return rowsAffected = 1

    End Function

    <System.ComponentModel.DataObjectMethod(System.ComponentModel.DataObjectMethodType.Delete, True)> _
    Public Function DeleteUserPermissions(ByVal userPermissionID As Integer) As Boolean

        Dim permissions As SPG.UserPermissionsDataTable = Adapter.GetUserPermissionByID(userPermissionID)
        Dim rowsAffected As Integer = 0

        If permissions.Count = 1 Then
            Dim userPermissions As SPG.UserPermissionsRow = CType(permissions.Rows(0), SPG.UserPermissionsRow)
            rowsAffected = Adapter.Delete(userPermissionID, userPermissions.ts)
        End If

        'Return true if precisely one row was deleted, otherwise return false.
        Return rowsAffected = 1

    End Function

    <System.ComponentModel.DataObjectMethod(System.ComponentModel.DataObjectMethodType.Delete, True)> _
    Public Function DeleteUserPermissions(ByVal username As String) As Boolean


        Dim rowsAffected As Integer = Adapter.DeletePermissionByUser(username)

        Return rowsAffected <> 0

    End Function

    Public Function GetPermissionObjects() As ArrayList

        Dim permissionObjects As New ArrayList
        Try
            Dim row As SPG.UserPermissionsRow
            For Each row In Me.Adapter.GetUserPermissions
                Dim permissionObject As String = row.PermissionObject
                If Not permissionObjects.Contains(permissionObject) Then
                    permissionObjects.Add(permissionObject)
                End If
            Next
        Finally

        End Try
        Return permissionObjects

    End Function

    Public Function CreateNewUserPermissions(ByVal UserName As String) As Boolean

        Return Convert.ToBoolean(Adapter.CreateNewUserPermissions(UserName))

    End Function

    

End Class

File: SuperiorPackGroup/BLL/UserPermissionsBLL.cs
//INSTANT C# NOTE: Formerly VB project-level imports:
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Drawing;
using System.Diagnostics;
using System.Windows.Forms;

using System;
using SuperiorPackGroup.SPGTableAdapters;
using System.Text;
using DXDAL.SPGData;
using System.Linq;

namespace SuperiorPackGroup
{
	[System.ComponentModel.DataObject()]
	public class UserPermissionsBLL
	{

		private UserPermissionsTableAdapter m_UserPermissionsTableAdapter = null;
		public UserPermissionsTableAdapter Adapter
		{

			get
			{
				if (m_UserPermissionsTableAdapter == null)
				{
					m_UserPermissionsTableAdapter = new UserPermissionsTableAdapter();
					m_UserPermissionsTableAdapter.Connection.ConnectionString = Properties.Settings.Default.UserConnectionString;
				}
				return m_UserPermissionsTableAdapter;
			}

		}

		[System.ComponentModel.DataObjectMethod(System.ComponentModel.DataObjectMethodType.Select, false)]
		public SPG.UserPermissionsDataTable GetUserPermissionsByUserName(string userName, string permissionObject)
		{

			return Adapter.GetUserPermissionsByUserName(userName, permissionObject);

		}

		private void UpdateAuditTrail(SPG.UserPermissionsRow ModifiedRecord, object[] originalRecord)
		{

			StringBuilder builder = new StringBuilder(string.Empty);
			int recordFields = (ModifiedRecord.ItemArray.Length - 1);
			int i = 0;
			while (i <= recordFields)
			{
				try
				{
					if (Convert.IsDBNull(originalRecord[i]))
					{
						if (!Convert.IsDBNull(ModifiedRecord[i]))
						{
							builder.Append(string.Format("{0}:{1}({2}); ", ModifiedRecord.Table.Columns[i].ColumnName, "NULL", ModifiedRecord[i]));
						}
					}
					else if (Convert.IsDBNull(ModifiedRecord[i]))
					{
						if (!Convert.IsDBNull(originalRecord[i]))
						{
							builder.Append(string.Format("{0}:{1}({2}); ", ModifiedRecord.Table.Columns[i].ColumnName, originalRecord[i], "NULL"));
						}
					}
					else if (ModifiedRecord[i] != originalRecord[i])
					{
						builder.Append(string.Format("{0}:{1}({2}); ", ModifiedRecord.Table.Columns[i].ColumnName, originalRecord[i], ModifiedRecord[i]));
					}
				}
				catch
				{
				}
				i += 1;
			}
			if (builder.Length > 2)
			{
				builder.Length = (builder.Length - 2);
				AuditTrailBLL.AddTrailEntry(Properties.Settings.Default.UserName, Convert.ToInt32(ModifiedRecord[0]), "UserPermissions", builder.ToString());
			}
		}

		[System.ComponentModel.DataObjectMethod(System.ComponentModel.DataObjectMethodType.Update, true)]
		public bool UpdateUserPermissions(int? UserPermissionID, string PermissionObject, string PermissionLevel, string UserName)
		{

			SPG.UserPermissionsDataTable permissions = Adapter.GetUserPermissionByID(UserPermissionID.Value);

			if (permissions.Count() == 0)
			{
				//It is a new record
				return InsertUserPermissions(PermissionObject, PermissionLevel, UserName);
			}

			SPG.UserPermissionsRow userPermissions = permissions[0];

			object[] originalRecord = userPermissions.ItemArray;

			userPermissions.PermissionObject = PermissionObject;
			userPermissions.PermissionLevel = PermissionLevel;
			userPermissions.UserName = UserName;

			int rowsAffected = Adapter.Update(userPermissions);

			if (!(originalRecord == null))
			{
				UpdateAuditTrail(userPermissions, originalRecord);
			}

			return rowsAffected == 1;

		}

		[System.ComponentModel.DataObjectMethod(System.ComponentModel.DataObjectMethodType.Insert, true)]
		public bool InsertUserPermissions(string PermissionObject, string PermissionLevel, string UserName)
		{

			SPG.UserPermissionsDataTable permissions = new SPG.UserPermissionsDataTable();
			SPG.UserPermissionsRow userPermissions = permissions.NewUserPermissionsRow();

			userPermissions.UserName = UserName;
			userPermissions.PermissionObject = PermissionObject;
			userPermissions.PermissionLevel = PermissionLevel;

			permissions.AddUserPermissionsRow(userPermissions);
			int rowsAffected = Adapter.Update(permissions);

			return rowsAffected == 1;

		}

		[System.ComponentModel.DataObjectMethod(System.ComponentModel.DataObjectMethodType.Delete, true)]
		public bool DeleteUserPermissions(int userPermissionID)
		{

			SPG.UserPermissionsDataTable permissions = Adapter.GetUserPermissionByID(userPermissionID);
			int rowsAffected = 0;

			if (permissions.Count() == 1)
			{
				SPG.UserPermissionsRow userPermissions = (SPG.UserPermissionsRow)permissions.Rows[0];
				rowsAffected = Adapter.Delete(userPermissionID, userPermissions.ts);
			}

			//Return true if precisely one row was deleted, otherwise return false.
			return rowsAffected == 1;

		}

		[System.ComponentModel.DataObjectMethod(System.ComponentModel.DataObjectMethodType.Delete, true)]
		public bool DeleteUserPermissions(string username)
		{


			int rowsAffected = Adapter.DeletePermissionByUser(username);

			return rowsAffected != 0;

		}

		public ArrayList GetPermissionObjects()
		{

			ArrayList permissionObjects = new ArrayList();
			try
			{
//INSTANT C# NOTE: Commented this declaration since looping variables in 'foreach' loops are declared in the 'foreach' header in C#:
//				Dim row As SPG.UserPermissionsRow
				foreach (SPG.UserPermissionsRow row in Adapter.GetUserPermissions())
				{
					string permissionObject = row.PermissionObject;
					if (!permissionObjects.Contains(permissionObject))
					{
						permissionObjects.Add(permissionObject);
					}
				}
			}
			finally
			{

			}
			return permissionObjects;

		}

		public bool CreateNewUserPermissions(string UserName)
		{

			return Convert.ToBoolean(Adapter.CreateNewUserPermissions(UserName));

		}



	}

}